### ISSUE TITLE
Feature Request - Add DirectInput  - Support

### ISSUE THREAD

**Issue by h1cks** [amerkoleci/Vortice.Windows#167]
HI, expect that this may not go too far.   

But one of the challenges I have is the current XInput is not fully supportive of non XBox Controllers which leaves a gap

Good work on the rest btw.  Just a query.

**Comment by amerkoleci** [comment]
DirectInput? DirectX11 support?

**Comment by h1cks** [comment]
> DirectInput? DirectX11 support?

Mis type. I was thinking about asking media engine support but changed the thread.  Didn't update fully.   Looking at the media foundation but having trouble looking for samples

**Comment by amerkoleci** [comment]
Once I'm back from vacation, August 23th I'll start adding DirectInput support

**Comment by h1cks** [comment]
Thanks.  Have a great holiday!

**Comment by BMarques** [comment]
This is especially important to support steering wheels' force feedback as well.

**Comment by amerkoleci** [comment]
@BMarques as specified any feature request must be sponsored, I know lot of features are important but this is spare time project, @h1cks is sponsoring me for long time already

**Comment by BMarques** [comment]
Ah ok, but I didn't mean it in that way :). Just referencing as well another reason why DInput is still valid these days :).

**Comment by amerkoleci** [comment]
Initial bindings just landed with commit: https://github.com/amerkoleci/Vortice.Windows/commit/4a008394a7568dbc1484e6ae7c4e2f936d43d5c6

Will improve it further but you can start playing with it.

**Comment by h1cks** [comment]
@amerkoleci  thanks, I will start digging into it after I finish my volumetric rendering work :)

**Comment by amerkoleci** [comment]
DirectInput feels almost completed to me, please test and as I plan to push nuget release by the end of the week.

**Comment by h1cks** [comment]
@amerkoleci I will test, just merging my latest sharpdx updates into my vortice version.   I'm also testing some issues with 3d sound, I use so much of the dx11 interface it's a good stress test.  

if interested, I can share my game code base with you if you ever need to debug/test features.

will keep you informed of my progress with respects to Directinput testing.

**Comment by h1cks** [comment]
@amerkoleci did some testing on directinput but not sure if just my implementation is wrong here.



```
  _directInput = DInput.DirectInput8Create();
```

then enumeratealldevices which is successful

```
        static public Dictionary<Guid, DeviceInstance> EnumerateAllAttachedDevices(IDirectInput8 directInput)
        {
            Dictionary<Guid, DeviceInstance> connectedDeviceList_ = new Dictionary<Guid, DeviceInstance>();

            foreach (DeviceInstance deviceInstance_ in directInput.GetDevices(DeviceClass.Keyboard, DeviceEnumerationFlags.AttachedOnly))
            {
                if (deviceInstance_.Type == Vortice.DirectInput.DeviceType.Keyboard)
                {
                    connectedDeviceList_.Add(deviceInstance_.InstanceGuid, deviceInstance_);
                }
                else
                {
                    ErrorHandler.WriteConsoleLine(deviceInstance_.ProductName + " does not match input type, ignored.");
                }
            }

            return connectedDeviceList_;
        }
```

I think create the device with the guid.  

```

          _playerInput =  directInput.CreateDevice(guid);

                Debug.Assert(directInput.IsDeviceAttached(guid));
// Passes the IsDeviceattached check

                 Result result_ = _playerInput.Acquire();
// fails here.

                var state_ = _playerInput.GetCurrentKeyboardState();

```

Did this with keyboard and joystick, same result.   It maybe something that I am overlooking, but I believe this is the bare bones implementation.

**Comment by amerkoleci** [comment]
Once you created device you have to call SetDataFormat https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.DirectInput/IDirectInputDevice8.cs#L424, just like you would do when writing native code :

SetDataFormat<RawKeyboardState> for keyboard, etc

**Comment by h1cks** [comment]
Yes, but I am unable to find the presets of c_dfDIKeyboard

https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ee417925(v=vs.85)

also

c_dfDIMouse
c_dfDIMouse2
c_dfDIJoystick
c_dfDIJoystick2

which are part of the predefined formats.

I used 

```
  _joystickInput.SetDataFormat<RawJoystickState>();
```

just for a test, because it's taking a TRaw object in.

**Comment by amerkoleci** [comment]
Not sure how this scenario can be handled:
@andrew-boyarshin dinput.h has:

extern const DIDATAFORMAT c_dfDIMouse; 

Is it possible to handle this scenario with SharpGen?

**Comment by andrew-boyarshin** [comment]
@amerkoleci no, SharpGen doesn't do that. There is no .NET'y way to handle that. You have to use `GetProcAddress` if there is no wrapping getter/setter function.

**Comment by amerkoleci** [comment]
@andrew-boyarshin GetProcAddress won't resolve the problem, as c_dfDIMouse is not visible like function.

**Comment by h1cks** [comment]
Probably make a fool of myself here....

From what I can see from Sharpdx, there seems to be an custom classes for each device supported by a customer device class underlying that creates a dataformat object from all available attributes based upon the type of device.   But I don't have the full story, so i'm probably guessing a bit here.

https://github.com/sharpdx/SharpDX/blob/master/Source/SharpDX.DirectInput/CustomDevice.cs

When the class is instantiated, it seems to go call a GetDataFormat function that instantiates the dataformat based upon the type TRaw struct passed in.  Builds a dictionary, then once done builds the Dataformatobject, I assume this is an "almanac" or everything object.  the object returned is then used in SetDataFormat.

**Comment by h1cks** [comment]
IF what I have said above holds, it might just need some additional logic to generate a "everything" for device Dataformat object.  At the moment, the DataFormat object in Vortice is internal, so I can't reference it atm.  I am assuming I can build something similar.

**Comment by amerkoleci** [comment]
Right, if you call SetDataFormat<RawKeyboardState>() should initialize the keyboard, SetDataFormat<RawMouseState>()  the mouse etc

**Comment by h1cks** [comment]
> Right, if you call SetDataFormat() should initialize the keyboard, SetDataFormat() the mouse etc

Yes, but this code will need to be developed though.

**Comment by amerkoleci** [comment]
Feel free to submit PR, when I have some spare time I will add different samples here: https://github.com/amerkoleci/Vortice.Windows.Samples

**Comment by h1cks** [comment]
@amerkoleci I've been able to find that I should not step through SetCooperative which was giving a fail.

The issue now I can replicate in Sharpdx and Vortice, so there is no issue with vortice!  But....With one question, there is accessor in the Sharpdx code that exposes the properties and I set the buffersize for between each GetBufferedData.

If the buffersize is 0, then GetBufferedData throws an exception.  The same exception is thrown in Sharpdx that vortice throws.  Once I solve how to set that buffersize.  I will submit my example code into a PR.

in sharpdx, the Property I set the
```
device.Properties.BufferSize = c_defaultBufferSize;
```

I assume that there is a way to do this in Vortice?
